### PR TITLE
ECE374 Course Offering Page

### PR DESCRIPTION
**Pull request by xavierrouth**
😸

### CODE REVIEW

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+## 374B vs 374A
+There are two versions of 374 offered at UIUC, section A and section B. They have different lectures, different discussions, different homeworks and exams, different grading. Most people consider the B section to be less difficult than the A section. One can check the [grade disparity](https://waf.cs.illinois.edu/discovery/grade_disparity_between_sections_at_uiuc/) to see the actual GPA differences. The B section lacks the emphasis on formal mathematical proof writing that the A section has. The A section and B section have different amounts of midterms and different grading policies. Read the [CS 374A page](../CS%20Course%20Offerings/ECE374A.md) as well to learn more about the differences.
```
Comment (lnguy7) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: Add this do the X vs Y section plz I wanna try to keep all the classes uniform in terms of headers

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
+# ECE374B
+## Introduction to Algorithms and Models of Computation
+
+ECE374B is one of two **independent** sections of CS/ECE374, Introduction to Algorithms and Models of Computation. It is unoficially the 'ECE' version of 374. It is considered by some to be easier than the A version.
```
Comment (lnguy7) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: I like this we should add this to every page (like a little overview at the top)

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+- Divide & Conquer
+- Graph Algorithms
+- Shortest Paths, MSTs
+- Dynamic Programming
+- Directed Acylic Graphs
+- Algorithmic Complexity Analysis
+#### Secret Third Thing
+- Reductions
+- Undecidability
+- NP, NP-Hardness, P vs NP
```
Comment (lnguy7) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: I like this formatting we should use it for the other classes

**Review by lnguy7 — CHANGES_REQUESTED** [review]
Small change for the X vs Y part

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md (hunk 2 above)
Comment (alex-m-zhang) [reply docs/Course Wiki/ECE Course Offerings/ECE374B.md]: "unofficially" is spelled incorrectly.
No need for "Introduction to Algorithms and Models of Computation" header.
Probably remove the sentence about one is easier than the other. Instead, just say "Notes on the difference between the two sections can be found here" and link the taking X vs. Y page.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md (hunk 3 above)
Comment (alex-m-zhang) [reply docs/Course Wiki/ECE Course Offerings/ECE374B.md]: Can you add something about how the content flows from the beginning to the end (i.e. what topics in the second thing are used in the third thing)?

Remove periods at the end of each bullet point.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+## Prerequisites
+- [CS 173](../CS%20Course%20Offerings/CS173.md)
+- [CS 225](../CS%20Course%20Offerings/CS225.md)
+
+There are two official prerequisites to ECE 374, CS 173 - Discrete Structures, and CS 225 - Data Structures and Algorithms. 374B expands on the theoretical computer science from CS173 and covers algorithms more in depth than CS225, without any of the programming. For part one of the course, you should remember induction, set notation, and FSMs from CS 173. Induction and recursion are fundamental to the course, so it is important to know them well. In the second portion of the course, there is a significant overlap with the graph algorithms taught in CS225.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: You can delete the first sentence.
Delete the space between "CS" and "173". Should be "CS173". Repeat for CS225 as well.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+## When to Take it
+ECE374 is a required course for all Computer Engineering and Computer Science majors. ECE 374 is not used as a prerequisite for any course in the ECE department, so for students not interested in higher level CS courses, there is little reason to rush it. However, due to the knowledge needed from CS 225 and CS 173, it might help to take it within a few semesters after finishing CS 225. On the other hand, taking this course early would help to prepare for Dynamic Programming interview questions. It is very common to encouter DP questions when applying for software engineering internships, and the HWs in ECE374 help prepare for them well. This course is fairly challenging so take caution if you plan to take this course with too many other technical courses.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: Delete the space between "ECE" and "374". Should be "ECE374". Apply this everywhere.
Second sentence: Replace first ECE374 with "It".
"Taking this course early would help prepare for software interview questions, particularly those related to Dynamic Programming".
A bit more detail is needed on how to rush this course (just mention that one should take 225 asap).

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+## Course Structure
+In terms of content the course is split into three sections, each with their own midterm. The first 5 or so weeks will be spent covering the basics of models of computation. The next 6ish weeks will cover design and analysis of algorithms. And the rest of the course will cover decidability, reductions, and P vs NP.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: Last sentence: Delete "And", start sentence with "The".

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+During the semester the course has 9-10 Homeworks, due every week besides weeks of midterms. You are allowed to work with other students in groups of three. Homeworks include problems that were covered in discussion, so if you go to discussion you know the answer to 1-2 problems on each HW. About 75% of the grade comes from the three midterms and the final, and the rest of the 25% from the HWs. You are allowed to drop the lowest exam, including the final. This means that if you do well on the first three exams, you can completely skip the final. There is also a provided cheat sheet for each exam. This makes memorizing theoretical definitions less important, but it also makes harder questions appear on the exam.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: First sentence: "besides the weeks of the midterms."
Last sentence: "but allows harder questions to appear on the exam".

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+The course has two components, discussions and lecture. Lectures are twice a week and run for 75 minutes. Discussions are also twice a week and 50 minutes.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: Move this to the top of the section.
Also expand a bit on what discussions are useful for and how they work (and if they are worth going to).

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+## Instructors
+In recent years (as of Fall 2023), Proffessor Nickvash Kani has taught the B section of the course.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: Professor is spelt wrong.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+During the semester the course has 9-10 Homeworks, due every week besides weeks of midterms. You are allowed to work with other students in groups of three. Homeworks include problems that were covered in discussion, so if you go to discussion you know the answer to 1-2 problems on each HW. About 75% of the grade comes from the three midterms and the final, and the rest of the 25% from the HWs. You are allowed to drop the lowest exam, including the final. This means that if you do well on the first three exams, you can completely skip the final. There is also a provided cheat sheet for each exam. This makes memorizing theoretical definitions less important, but it also makes harder questions appear on the exam.
+
+The course has two components, discussions and lecture. Lectures are twice a week and run for 75 minutes. Discussions are also twice a week and 50 minutes.
+
+## Instructors
+In recent years (as of Fall 2023), Proffessor Nickvash Kani has taught the B section of the course. 
+
+## Course Tips
+Go to discussions, make sure to write down the problems so you know them when they come up in the homeworks.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: "Go to discussions and make sure to write down"

Put another line between each sentence so it shows up in the Markdown as different paragraphs.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+Make sure to show your work on exams, there are decent amounts of partial credit offered.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: "Make sure to show your work on exams, as there are decent amounts..."

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+Find a group for homeworks, but don't split up the problems as HW problems sometimes show up on exams.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: A lot more expansion is necessary here:
"The best way to do homeworks within your groups is to do them separately, then meet up and discuss your individual solutions. This allows you to ... Typing up can then be distributed evenly. HW problems sometimes show up on exams, so doing hws separately guarantees that you know all of the problems..." etc.
Also move this to the top - it's the most important tip for this course.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+## Life After
+ECE 374 opens students up to most of the upper level courses in the CS department. The most obvious follow up would be CS 475 - Formal Models of Computation or CS 473 - Algorithms. Some other popular options include CS 421 or CS 426.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: Switch order of 475 and 473 (so its in numerical order).
Also link the pages (even if they don't exist yet - we will have people writing those pages).

Include names for 421 and 426.
Also explain why these are good followups, and which ones to take if you liked which section of the course (ex. 473 is good if you really like the algorithms portion of the course).

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+## Infamous Topics
+- Language Transformations: The most difficult topic from the first section. You must know it well because there will be a problem about it on MT1.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: How to learn it well?

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+- Dynamic Programming: Hard but good to know for getting a Software Engineering Internship.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: "Dynamic Programming: Good to know for getting a software engineering internship"

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md:64]: Does proving NP hardness (and which direction the proof goes) also fit in here? Students seem to struggle with that.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+## Resources
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md:65]: Expand a bit more on what one would find in each of these and why they're so useful.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
++ The official course website: [https://ecealgo.com/](https://ecealgo.com/)
++ Prof Jeff Erickson's textbook and notes: [https://jeffe.cs.illinois.edu/teaching/algorithms/](https://jeffe.cs.illinois.edu/teaching/algorithms/)
++ 374A section website: [https://courses.engr.illinois.edu/cs374al1/sp2023/](https://courses.engr.illinois.edu/cs374al1/sp2023/)
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: No need to link this.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
++ FSM designer: [https://madebyevan.com/fsm/](https://madebyevan.com/fsm/)
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: Explain that this is very useful for homeworks.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+
+## Life After
+ECE 374 opens students up to most of the upper level courses in the CS department. The most obvious follow up would be CS 475 - Formal Models of Computation or CS 473 - Algorithms. Some other popular options include CS 421 or CS 426.
+
+## Infamous Topics
+- Language Transformations: The most difficult topic from the first section. You must know it well because there will be a problem about it on MT1.
+- Dynamic Programming: Hard but good to know for getting a Software Engineering Internship.
+
+## Resources
++ The official course website: [https://ecealgo.com/](https://ecealgo.com/)
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md:66]: No need to link this either.

File: docs/Course Wiki/X versus Y/CS374A vs ECE374B.md [hunk 21]
```diff
@@ -0,0 +1,12 @@
+# CS374A vs ECE374B
+## Relevant Course Reviews:
+- [CS 374A](../CS%20Course%20Offerings/CS374A.md)
+- [ECE 374B](../ECE%20Course%20Offerings/ECE374B.md)
+
+There are two **idependent** versions of 374 offered at UIUC, section A and section B. Both fulfill the same requirements and work as prerequisites for the same classes. They do however have different lectures, different discussions, different homeworks and exams, and different grading. Both sections are crosslisted as CS/ECE. The A section is run by the CS department so we refer to it as CS 374A. The B section is taught by professors from the ECE department, so we refer to it as ECE 374B.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/X versus Y/CS374A vs ECE374B.md]: "independent" is spelled wrong.
Remove "do": "However, they have different lectures, different discussions, ..."

File: docs/Course Wiki/X versus Y/CS374A vs ECE374B.md [hunk 22]
```diff
@@ -0,0 +1,12 @@
+# CS374A vs ECE374B
+## Relevant Course Reviews:
+- [CS 374A](../CS%20Course%20Offerings/CS374A.md)
```
Comment (alex-m-zhang) [inline docs/Course Wiki/X versus Y/CS374A vs ECE374B.md]: Change to CS374A and ECE374B (remove space).

**Review by alex-m-zhang — CHANGES_REQUESTED** [review]

**Comment by alex-m-zhang** [comment]
Is it worth it to talk about how 374A grades harsher but forces you to learn more vs. 374B? Or is that too inflammatory for the wiki. @lnguy7

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+- Undecidability
+- NP, NP-Hardness, P vs NP
+
+In lecture it may seem that the algorithms covered aren't very connected to NP hardness or P vs NP. However on exams and homeworks (especially in the 3rd section onwards), you will be asked to classify new problems using the algorithms covered in the 2nd section as a starting point via reductions. 
+
+## Prerequisites
+- [CS173](../CS%20Course%20Offerings/CS173.md)
+- [CS225](../CS%20Course%20Offerings/CS225.md)
+
+ 374B expands on the theoretical computer science from CS173 and covers algorithms more in depth than CS225, without any of the programming. For part one of the course, you should remember induction, set notation, and FSMs from CS173. Induction and recursion are fundamental to the course, so it is important to know them well. In the second portion of the course, there is a significant overlap with the graph algorithms taught in CS225.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: First word: ECE374B.

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+
+## When to Take it
+ECE374 is a required course for all Computer Engineering and Computer Science majors. It is not used as a prerequisite for any course in the ECE department, so for students not interested in higher level CS courses, there is little reason to rush it. If you do decide you want to take ECE374 as early as possible, you need to take CS225 as early as possible, as well as finishing [Junior Eligibility](https://ece.illinois.edu/academics/ugrad/advising-tips/junior225) requirements. However, due to the knowledge needed from CS225 and CS173, it might help to take it within a few semesters after finishing CS225. On the other hand, Taking this course early would help prepare for software interview questions, particularly those related to Dynamic Programming. It is very common to encouter DP questions when applying for software engineering internships, and the HWs in ECE374 help prepare for them well. This course is fairly challenging so take caution if you plan to take this course with too many other technical courses.
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: "On the other hand, taking this course early" (uncapitalize the t in taking)

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+
+## Infamous Topics
+- Language Transformations: The most difficult topic from the first section. You must know it well because there will be a problem about it on MT1. TAs usually can explain language transformations well. There are also good examples in the textbook.
+- Dynamic Programming: Good to know for getting a Software Engineering Internship.
+- NP hardness: Proofs on NP-Hardness can get confusing quick. It is helpful to draw out diagrams in order to not get confused on which direction the proof goes.
+
+## Resources
++ The official course website: [https://ecealgo.com/](https://ecealgo.com/)
++ Prof Jeff Erickson's textbook and notes: [https://jeffe.cs.illinois.edu/teaching/algorithms/](https://jeffe.cs.illinois.edu/teaching/algorithms/)
++ 374A section website: [https://courses.engr.illinois.edu/cs374al1/sp2023/](https://courses.engr.illinois.edu/cs374al1/sp2023/)
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: I would change this to the FA23 version if you're going to leave it on here (although right now we do not know what that link is...)
Just delete it to save some hassle. You can keep the B-section link though.

**Review by alex-m-zhang — CHANGES_REQUESTED** [review]

File: docs/Course Wiki/ECE Course Offerings/ECE374B.md [hunk 26]
```diff
@@ -0,0 +1,69 @@
+# ECE374B
+ECE374B is one of two **independent** sections of CS/ECE374, Introduction to Algorithms and Models of Computation. It is unofficially the 'ECE' version of 374. Notes on the differences between the two sections can be found [here](../X%20versus%20Y/CS374A%20vs%20ECE374B.md).
```
Comment (alex-m-zhang) [inline docs/Course Wiki/ECE Course Offerings/ECE374B.md]: Change to add more detail, like so:
ECE374B is one of two **independent** sections of CS/ECE374, Introduction to Algorithms and Models of Computation. It is unofficially the 'ECE' version of 374. Notes on the differences between the two sections can be found [here](../X%20versus%20Y/CS374A%20vs%20ECE374B.md). It is a 4-credit hour course that is specifically required for CEs and qualifies as a technical elective for EEs. It is offered in the fall and spring.

**Review by alex-m-zhang — CHANGES_REQUESTED** [review]
Oops, one more change. Sorry

**Review by alex-m-zhang — APPROVED** [review]
LGTM!

**Comment by lnguy7** [comment]
> Is it worth it to talk about how 374A grades harsher but forces you to learn more vs. 374B? Or is that too inflammatory for the wiki. @lnguy7

We should add this to the ECE374B vs. CS374A section of the wiki

**Review by lnguy7 — APPROVED** [review]
LGTM

**Comment by lnguy7** [comment]
@alex-m-zhang Am merging

**Pull request merged**